Encode repeating

shrutivanw opened the pull request:
CS Fundamentals' assignment to replace repeating characters in the input string with character followed by character count.

shrutivanw committed:
Encode repeating

CS Fundamentals' assignment to replace repeating characters in the input string with character followed by character count.

Hamled commented on README.md:
```diff
@@ -1 +1,12 @@
-# encode_repeating
\ No newline at end of file
+# Encode Repeating
```
Should we link them to the [Wikipedia article for run-length encoding](https://en.wikipedia.org/wiki/Run-length_encoding) for more context around this idea?

shrutivanw replied:
That's a great link! Let me add it as a further reading link.

Hamled commented on README.md:
```diff
@@ -1 +1,12 @@
+   - "ddbbfffgjjjj" will get replaced with "ddbbf3gj4". "dd" and "bb" do not get replaced since the length will remain the same with "d1" and "b1". "g" does not get replaced with "g1" since that will increase the length.
+* Share and explain the time and space complexities for your solution.
+* At minimum, your implementation should pass the basic tests.
+**Note**: Do not use Ruby provided functionality for `.reverse` and `.reverse!`. You may use `.length` and `.slice!` to implement this method.
```
We should have an extra blank line above this so that the note goes into its own paragraph.

shrutivanw replied:
Done!

Hamled commented on lib/encode_repeating.rb:
```diff
@@ -0,0 +1,6 @@
+# with a number representing the frequency. The replacement is done only if the
+# string length will get reduced by this process.
+def encode_repeating(my_string)
+  my_string << "NOT IMPLEMENTED"
```
How would you feel about doing `raise NotImplementedError` here instead?

Best practice I think is generally to avoid `puts` statements and other things that will print to the console for code being tested, because that output kind of gets randomly interspersed with the test runner's output.

shrutivanw replied:
Great idea!

Hamled commented on specs/encode_repeating_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+describe "encode repeating" do
+  describe "basic tests" do
+    describe "test 1" do
```
Same as the other repos, we should probably remove the `describe` blocks that are immediate parents of individual tests.

Hamled commented on specs/encode_repeating_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+  # if it's a string parameter, check for empty
+  describe "empty string input test" do
+    it "empty string remains as is" do
```
I think we could move this test and the nil test into the edge cases section.

Hamled commented on specs/encode_repeating_spec.rb:
```diff
@@ -0,0 +1,82 @@
+  describe "edge cases" do
+    describe "edge case 1" do
+      it "string with no consecutive repeating characters" do
+        test_string = "at53fag"
```
I think this is a good edge case test, however I have two concerns:
1. Do we have specific instructions for when the input string includes digits? I think the simplest thing would be to have no tests with digits, or to expect that `ArgumentError` is raised when the input string has digits.
1. I'm sure it was unintentional, but we should remove the slur from this test string.

shrutivanw replied:
Whoops! Good catch on the slur - that didn't even occur to me as typed random letters. Oh boy! Thank you!
I changed the test string for the no consecutive repeating character to be "aeiou"

shrutivanw committed:
Incorporating feedback

- Making the readme file more readable and including link to wikipedia for RLE.
- Moving nil and empty test cases to edge cases.
- Fixing the non repeating characters test case
- Making the method template raise NotImplementedError

shrutivanw commented:
Thank you @Hamled PR updated.
